### ISSUE TITLE
[bug] Issue with working directory in set_version

### ISSUE THREAD

**Issue by jasal82** [conan-io/conan#6129]
We have problems with the new `set_version()` method. Up to now we assigned the result of a function call directly to the `version` member. Now we tried to switch to the `set_version()` approach, but there seems to be a difference between the working directories as shown in the following two examples:

```python
from conans import ConanFile, tools
import os

def get_version():
  print(os.getcwd())

class Package(ConanFile):
  name = "test"
  version = get_version()
```

When exporting the recipe above (`mkdir build && cd build && conan export .. asaljo/testing`) the current working directory is set to the source folder as expected. Now compare that with the following:

```python
from conans import ConanFile, tools
import os

class Package(ConanFile):
  name = "test"
  def set_version(self):
    print(os.getcwd())
```

Here the working directory is set to the build folder instead! We must read the contents of a header file from the source directory in this step (using `tools.load()`), but there is no way to do this in the `set_version()` approach, because there's also no `self.source_folder` property in that context.

**Comment by jasal82** [comment]
There's another problem with deriving version names from function calls which might be related. I tried to move the function that parses the version header into a base conanfile like this:

```python
class PythonRequires(ConanFile):
  name = "bundle_conanfile"
  version = "1.0"

def get_bundle_version():
  # parse version
  return version
```

I use that from the actual package conanfile like this:

```python
base_recipe = python_requires("bundle_conanfile/1.0@asaljo/testing")

class Builder(base_recipe.get_conanfile()):
  name = "test"
  version = base_recipe.get_bundle_version()
```

Note that I still use the `version =` approach because of the beforementioned bug. Now this works fine when exporting the recipe and package. But as soon as I consume that package from another one the `conan install` will fail because in that case the working directory seems to be set to `/home/asaljo/.conan/data/PACKAGE/VERSION/USER/CHANNEL/export`.

**Comment by memsharded** [comment]
I have just submitted a PR to define the conanfile.py folder, not the current folder. I had some concern, but your python_require comment seems an extra argument in favor of this PR: https://github.com/conan-io/conan/pull/6130.

The ``set_version()`` approach should work with ``python-requires``, at least with the latest improvements we are preparing (hopefully for Conan 1.21 too) in https://github.com/conan-io/conan/pull/5804/files.

**Comment by jasal82** [comment]
Great. I can apply the patch manually and check if it fixes my issues if you want.

**Comment by memsharded** [comment]
@jasal82 yes, testing this PR would be great. No, #5804 is a new approach for python_requires, it is not necessary, I have just checked, the reuse of ``set_version()`` already works with the current python-requires, there is a test for it too. Thanks!

**Comment by jasal82** [comment]
@memsharded First issue is fixed by the PR, second one is also gone, probably because I can use `set_version()` now instead.

**Comment by jasal82** [comment]
I just read through the comments in the PR. I'm not sure if you are aware of the fact that I actually need a reference to the source folder, not the folder where the conanfile is located. Thus passing `os.path.dirname(__file__)` does not help since it resolves to the `export` directory when run from the cache.

**Comment by memsharded** [comment]
Hi @jasal82 

Not sure what you mean here. The ``set_version()`` is never evaluated in the Conan cache, it is only evaluated at "export" time, and it evaluates in the user folder. Thus, ``__file__`` is pointing to the conanfile.py in the user space, and the files to get the version should be located there. Could you please clarify? Thanks!

**Comment by jasal82** [comment]
The thing is I get an exception from my `get_bundle_version()` function during `conan install` (as I wrote above). So it seems that it is run somehow during the install phase as well.

**Comment by memsharded** [comment]
Uhmmm, I am checking the code, and trying to reproduce, and I don't see how it would be called during ``install``, if it is in ``set_version()`` only, even with the ``python_requires``. Are you sure you are not doing ``version = base_recipe.get_bundle_version()``?

I am available in the CppLang slack too, if you want to check there.

**Comment by jasal82** [comment]
Ah forget it. I mixed up the contexts. The exception during the install phase is coming from the `set_version()` of the consumer. Both are using the same `python_requires` so it was not easy to tell which one was causing it. Sorry for the confusion.

**Comment by memsharded** [comment]
Fixed, now explicit use of ``self.recipe_folder`` will help to define the working directory correctly. Will be released next Conan 1.21
